subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskOutputs.java

```diff
@@ -87,14 +91,37 @@ public void run() {
 
     @Override
     public boolean isCacheEnabled() {
-        return !cacheIfSpec.getSpecs().isEmpty() && cacheIfSpec.isSatisfiedBy(task)
-            && (doNotCacheIfSpec.isEmpty() || !doNotCacheIfSpec.isSatisfiedBy(task));
+        return isCacheEnabled(new ArrayList<TaskCachingDisabledReason>());
     }
 
     @Override
-    public boolean isCacheAllowed() {
+    public boolean isCacheEnabled(List<TaskCachingDisabledReason> reasons) {
```

Comment:
Should we do something to cache the result of this? i.e., so we don't re-evaluate the Specs multiple times?